### PR TITLE
Add support for user-defined rules

### PR DESCRIPTION
**Pull request by vchuravy**
TODO:
- [ ] Docs and examples
- [ ] CLeanup interface a bit
- [ ] Rewrite internal rules to match

### CODE REVIEW

**Comment by zsz00** [comment]
What is the development status of this now?

**Comment by vchuravy** [comment]
Pending a design decision on #172

**Comment by wsmoses** [comment]
API is still in need of work and right now only adds custom rules for forward mode.

But nevertheless it does work, and has (and passes) tests, so I want to merge.

File: src/compiler/interpreter.jl [hunk 1]
```diff
@@ -139,9 +152,22 @@ enzyme_inlining_policy(@nospecialize(src)) = Core.Compiler.default_inlining_poli
...
 function Core.Compiler.resolve_todo(todo::InliningTodo, state::InliningState{S, T, <:typeof(enzyme_inlining_policy)}) where {S<:Union{Nothing, Core.Compiler.EdgeTracker}, T}
     mi = todo.mi
-    if is_primitive_func(mi.specTypes)
+    if is_primitive_func(mi.specTypes) 
+        return Core.Compiler.compileable_specialization(state.et, todo.spec.match)
+    end
+    if EnzymeRules.has_frule(mi.specTypes) || EnzymeRules.has_rrule(mi.specTypes)
         return Core.Compiler.compileable_specialization(state.et, todo.spec.match)
     end
+    ## TODO @vchuravy separate by mode like below
```
Comment (wsmoses) [inline src/compiler/interpreter.jl:161]: @vchuravy how might I extract the interp (and thus mode) here?

File: test/rules.jl [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    rule_f(x) = x^2
+
+    function rule_fip(x)
+       x[1] *= x[1]
+       return nothing
+    end
+
+    function EnzymeRules.forward(::Type{Tuple{typeof(rule_f), Float64}}, RT, Args)
+        @assert Args[1] <: Const
```
Comment (wsmoses) [inline test/rules.jl:15]: okay we still need to pass width explicitly as this could take no args and allocate

**Comment by wsmoses** [comment]
The 1.6 issue seems like an inling issue?

**Comment by vchuravy** [comment]
Vetoing merging this for now, we should first release a version with the accumulated fixes before landing this PR.

**Comment by wsmoses** [comment]
@vchuravy with the landed jll bump, have time to review this (and possibly ponder 1.6 no inling)

**Comment by vchuravy** [comment]
I have a strong preference to focus on stabilizing Enzyme.jl for a new release, a stretchgoal for that is GC integration and visit this for the next minor bump of Enzyme.jl
I am not convinced that we want to commit to the interface yet and I would like to see the current internal custom rules be ported over to this format so that we can evaluate whether or not it is the right approach,

File: src/compiler/interpreter.jl (hunk 1 above)
Comment (vchuravy) [reply src/compiler/interpreter.jl:168]: Maybe @aviatesk knows of the top of his head

Comment (aviatesk) [reply src/compiler/interpreter.jl:168]: I think it's `state.interp`.

File: test/rules.jl [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+@testset "Custom Forward Rules" begin
+
+    rule_f(x) = x^2
+
+    function rule_fip(x)
+       x[1] *= x[1]
+       return nothing
+    end
+
+    function forward(::Type{Tuple{typeof(rule_f), Float64}}, rt, args)
```
Comment (vchuravy) [inline test/rules.jl:16]: This seems a bit awkward. It really doesn't jive with multiple dispatch that we have to do subtype queries inside the rule.

```
function forward(::Const{typeof(rule_f)}, RT::Type{<:DuplicatedNoNeed}, x::Duplicated)
    return 10+2*x.val*x.dval
end

function forward(::Const{typeof(rule_f)}, RT::Type{<:BatchDuplicatedNoNeed}, x::BatchDuplicated{T, N}) where {T, N}
    return NTuple{N, T}(1000+2*x.val*dv for dv in x.dval)
end

function forward(::Const{typeof(rule_f)}, RT::Type{Duplicated}, x::Duplicated)
    return Duplicated(func.val(x.val), 100+2*x.val*x.dval)
end

function forward(::Const{typeof(rule_f)}, RT::Type{DuplicatedNoNeed}, x::BatchDuplicated{T, N}) where {T,N}
    return BatchDuplicated(func.val(x.val), NTuple{N, T}(10000+2*x.val*dv for dv in x.dval))
end
```

Would seem more julian to me.

**Comment by vchuravy** [comment]
Haven't thought through reverse mode yet, but I think:

```
using EnzymeCore
import EnzymeCore.EnzymeRules: forward

f(x) = x^2

function forward(::Const{typeof(f)}, RT::Type{<:DuplicatedNoNeed}, x::Duplicated)
    return 10+2*x.val*x.dval
end

function forward(::Const{typeof(f)}, RT::Type{<:BatchDuplicatedNoNeed}, x::BatchDuplicated{T, N}) where {T, N}
    return NTuple{N, T}(1000+2*x.val*dv for dv in x.dval)
end

function forward(::Const{typeof(f)}, RT::Type{<:Duplicated}, x::Duplicated)
    return Duplicated(func.val(x.val), 100+2*x.val*x.dval)
end

function forward(::Const{typeof(f)}, RT::Type{<:BatchDuplicated}, x::BatchDuplicated{T, N}) where {T,N}
    return BatchDuplicated(func.val(x.val), NTuple{N, T}(10000+2*x.val*dv for dv in x.dval))
end

import EnzymeCore: Annotation
function has_frule(@nospecialize(f))
    TT = Tuple{<:Annotation{Core.Typeof(f)}, Type{<:Annotation}, Vararg{<:Annotation}}
    isapplicable(forward, TT)
end

# Do we need this one?
function has_frule(@nospecialize(f), @nospecialize(RT))
    TT = Tuple{<:Annotation{Core.Typeof(f)}, Type{RT}, Vararg{<:Annotation}}
    isapplicable(forward, TT)
end

# Do we need this one?
function has_frule(@nospecialize(f), @nospecialize(RT), @nospecialize(TT))
    TT = Base.unwrap_unionall(TT)
    TT = Tuple{<:Annotation{Core.Typeof(f)}, Type{RT}, TT.parameters...}
    isapplicable(forward, TT)
end

# Base.hasmethod is a precise match we want the broader query.
function isapplicable(@nospecialize(f), @nospecialize(TT); world=Base.get_world_counter())
    tt = Base.to_tuple_type(TT)
    sig = Base.signature_type(f, tt)
    return !isempty(Base._methods_by_ftype(sig, -1, world)) # TODO cheaper way of querying?
end

@assert has_frule(f)
@assert has_frule(f, Duplicated)
@assert has_frule(f, DuplicatedNoNeed)
@assert has_frule(f, BatchDuplicated)
@assert has_frule(f, BatchDuplicatedNoNeed)
@assert has_frule(f, Duplicated, Tuple{<:Duplicated})
@assert has_frule(f, DuplicatedNoNeed, Tuple{<:Duplicated})
@assert has_frule(f, BatchDuplicated, Tuple{<:BatchDuplicated})
@assert has_frule(f, BatchDuplicatedNoNeed, Tuple{<:BatchDuplicated})

@assert !has_frule(f, Duplicated, Tuple{<:BatchDuplicated})
@assert !has_frule(f, DuplicatedNoNeed, Tuple{<:BatchDuplicated})
@assert !has_frule(f, BatchDuplicated, Tuple{<:Duplicated})
@assert !has_frule(f, BatchDuplicatedNoNeed, Tuple{<:Duplicated})
```

Is a decent start for the forward API.

**Comment by vchuravy** [comment]
Rebased in #589

**Pull request closed without merging**